crates/accelerate/src/synthesis/evolution/mod.rs

```diff
@@ -0,0 +1,439 @@
+
+use rustworkx_core::petgraph::graph::NodeIndex;
+use rustworkx_core::petgraph::prelude::StableDiGraph;
+use rustworkx_core::petgraph::Incoming;
+
+/// A Qiskit gate. The quantum circuit data returned by the pauli network
+/// synthesis algorithm will consist of clifford and rotation gates.
+type QiskitGate = (StandardGate, SmallVec<[Param; 3]>, SmallVec<[Qubit; 2]>);
+
+/// Expands the sparse pauli string representation to the full representation.
+///
+/// For example: for the input `sparse_pauli = "XY", qubits = [1, 3], num_qubits = 6`,
+/// the function returns `"IXIYII"`.
+fn expand_pauli(sparse_pauli: String, qubits: Vec<u32>, num_qubits: usize) -> String {
+    let mut v: Vec<char> = vec!['I'; num_qubits];
+    for (i, q) in qubits.iter().enumerate() {
+        v[*q as usize] = sparse_pauli.chars().nth(i).unwrap();
+    }
+    v.into_iter().collect()
+}
+
+/// Return the Qiskit's gate corresponding to the given Rustiq's Clifford gate.
+fn qiskit_clifford_gate(rustiq_gate: &CliffordGate) -> QiskitGate {
+    match rustiq_gate {
+        CliffordGate::CNOT(i, j) => (
+            StandardGate::CXGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32), Qubit(*j as u32)],
+        ),
+        CliffordGate::CZ(i, j) => (
+            StandardGate::CZGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32), Qubit(*j as u32)],
+        ),
+        CliffordGate::H(i) => (
+            StandardGate::HGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+        CliffordGate::S(i) => (
+            StandardGate::SGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+        CliffordGate::Sd(i) => (
+            StandardGate::SdgGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+        CliffordGate::SqrtX(i) => (
+            StandardGate::SXGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+        CliffordGate::SqrtXd(i) => (
+            StandardGate::SXdgGate,
+            smallvec![],
+            smallvec![Qubit(*i as u32)],
+        ),
+    }
+}
+
+/// Return the Qiskit rotation gate corresponding to the single-qubit Pauli rotation.
+///
+/// # Arguments
+///
+/// * py: a GIL handle, needed to negate rotation parameters in Python space.
+/// * paulis: Rustiq's data structure storing pauli rotations.
+/// * i: index of the single-qubit Pauli rotation.
+/// * angle: - Qiskit's rotation angle.
+fn qiskit_rotation_gate(py: Python, paulis: &PauliSet, i: usize, angle: &Param) -> QiskitGate {
+    let (phase, pauli_str) = paulis.get(i);
+    for (q, c) in pauli_str.chars().enumerate() {
+        if c != 'I' {
+            let standard_gate = match c {
+                'X' => StandardGate::RXGate,
+                'Y' => StandardGate::RYGate,
+                'Z' => StandardGate::RZGate,
+                _ => panic!(),
```

Comment:
Personally, I would use `unreachable` here as we don't expect this arm can be reached, whereas `panic` is maybe less precise? Plus it might be nice adding a small message to simplify tracking down where this came from 🙂 But I'm also happy to leave as is, this is just a suggestion.
```suggestion
                _ => unreachable!("Only X, Y and Z are possible characters at this point."),
```

Reply:
Changed (the commit will soon follow).